### ISSUE TITLE
*OrderByWithAggregationInput is not typesafe and can result in a PrismaClientValidationError

### ISSUE THREAD

**Issue by rivatove** [prisma/prisma#25854]
### Bug description

This is the type definition of the sorting object in Prisma.

```ts
  type tableOrderByWithAggregationInput = {
    A?: SortOrder
    B?: SortOrder
    C?: SortOrder
    ...
   }
```

Query builder accepts `tableOrderByWithAggregationInput | tableOrderByWithAggregationInput[]`. We are supposed to use an array when we are sorting by multiple columns. The issue is, specifying multiple columns to sort by in a single object will not result in a compiler error. But it will error at runtime.

Hence, I think the object's type should be rearchitected. I suppose a discriminated union would solve the issue.

### How to reproduce

Reproducable anywhere.

### Prisma information

-

### Environment & setup

-

### Prisma Version

6.0.1

**Comment by ludralph** [comment]
Hi @rivatove 

In the design of `orderBy` types, Prisma allows `tableOrderByWithAggregationInput | tableOrderByWithAggregationInput[]` for sorting. At runtime, multiple fields must be specified as an array of objects. However, the type definition does not enforce this restriction, leading to possible runtime errors. To avoid runtime issues, you can use [Prisma.validator](https://www.prisma.io/docs/orm/prisma-client/type-safety/prisma-validator#using-the-prismavalidator) to enforce correct input types at compile time:
```js
const orderBy = Prisma.validator<Prisma.UserFindManyArgs>()({
  orderBy: [
    { A: 'asc' },
    { B: 'desc' }
  ]
});
```
Your suggestion of using a discriminated union could be a potential solution. You can open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that our engineers can evaluate it.

**Comment by rivatove** [comment]
Hi @ludralph,

Using Prisma.validator is not a systemic solution to the issue. I'm not looking to validate a reusable part of a query. I want individual usages of the orderBy field be typesafe.

I disagree this is a feature request. It's a bug, Prisma is not typesafe under the reported circumstances.

**Comment by aqrln** [comment]
Yeah, this absolutely sounds like a bug. I don't see how a discriminated union would solve the issue but an easy an backward compatible fix would be making `tableOrderByWithAggregationInput` a union of `{ A: SortOrder } | { B: SortOrder } | ...` and accepting `tableOrderByWithAggregationInput | tableOrderByWithAggregationInput[] | Record<string | never>`. An empty object case must be handled separately because the current type definition also accepts something like `[{}, {}, {}]` which makes no sense.

**Comment by rivatove** [comment]
> Yeah, this absolutely sounds like a bug. I don't see how a discriminated union would solve the issue but a simple and backward compatible fix would be making `tableOrderByWithAggregationInput` a union of `{ A: SortOrder } | { B: SortOrder } | ...` and accepting `tableOrderByWithAggregationInput | tableOrderByWithAggregationInput[] | Record<string | never>`. An empty object case must be handled separately because the current type definition also accepts something like `[{}, {}, {}]` which makes no sense.

Yes, that's what I meant. And the `Record<string, never>` could get stripped when a release allows for breaking change.

**Comment by dcodesdev** [comment]
I don't see why an array must be provided if there are more than one order values, using an object with multiple keys should still work and it can be turned back into an array behind the scenes achieving the same functionality, that way it will be type safe and backwards compatible too.

**Comment by Ryanjso** [comment]
I get no type error for the following line in a findMany
`orderBy: [{ last: "asc", first: "asc" }],`

however this throws a runtime error
`Argument "orderBy": Invalid value provided. Expected UserOrderByWithRelationInput, provided (Object).`

Using Prisma 6.7
